ui: fix cursor problems

erw7 opened the pull request:
- Change to reset everything before setting `guicursor`.
- tui: Fix a case where the cursor was not displayed after hiding the   cursor and then setting it to be displayed again.

fixes #12800.

### Steps to reproduce using `nvim -u NORC`

```vim
nvim -u NORC
:set termguicolors
:hi nCursor guifg=red guibg=red
:hi iCursor guifg=green guibg=green
:hi cCursor guifg=blue guibg=blue
:set guicursor=n:block-nCursor,i:hor25-iCursor,c:ver25-cCursor
:set guicursor-=c:ver25-cCursor 
:
```

### Actual behaviour

The cursor is a blue vertical.

### Expected behaviour

The cursor should be the default color block.

- [x] Document.
- [x] Test.

erw7 committed:
ui: fix cursor problems

- Change to reset everything before setting guicursor.
- tui: Fix a case where the cursor was not displayed after hiding the
  cursor and then setting it to be displayed again.

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -99,11 +99,23 @@ char_u *parse_shape_opt(int what)
   for (round = 1; round <= 2; round++) {
+    // Set the defaults, for the missing parts
+    if (round == 2 || *p_guicursor == NUL) {
+      for (idx = 0; idx < SHAPE_IDX_COUNT; idx++) {
```
this for loop should go in a `static clear_shap_table()` function. That avoids setting `idx` here which is fragile, also helps readability.

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -99,11 +99,23 @@ char_u *parse_shape_opt(int what)
 
   // First round: check for errors; second round: do it for real.
   for (round = 1; round <= 2; round++) {
+    // Set the defaults, for the missing parts
```
"for the missing parts" is no longer relevant, right?

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -99,11 +99,23 @@ char_u *parse_shape_opt(int what)
 
```
let's mention this in the docstring for `parse_shape_opt`:

```
/// Parses the 'guicursor' option, or clears `shape_table` if 'guicursor' is empty.
///
/// @param ...

justinmk reviewed (approved):
looks like a good idea, definitely worth testing but can probably just test it by checking 'guicursor' before/after, don't need a TUI test.

erw7 committed:
doc: fix the description of the `guicursor`

erw7 committed:
test: fix cursor_spec

Changed the setting to be cleared when setting the `guicursor`, so the test
is changed accordingly.

justinmk commented on test/functional/ui/cursor_spec.lua:
```diff
@@ -286,6 +286,21 @@ describe('ui/cursor', function()
+    meths.set_option('guicursor', 'n:ver35-blinkwait171-blinkoff172-blinkon173-Cursor/lCursor')
+    screen:expect(function()
+      for _,m in ipairs(screen._mode_info) do
+        if m.name ~= 'normal' then
```
should we explicitly test the "normal" case also?

erw7 replied:
I think the case where the setting exists in `guicursor` has been well tested in previous tests. Therefore, I don't think it's necessary.

teto commented on src/nvim/cursor_shape.c:
```diff
@@ -77,7 +81,8 @@ Array mode_style_array(void)
 }
 
-/// Parse the 'guicursor' option
+/// Parse the 'guicursor' option, or clears `shape_table` if 'guicursor'
```
I think doxygen allows to reference symbols via `\ref` so you could write `\ref  shape_table`.

teto reviewed (commented):
@lambdalisue does that fix it for you ?

justinmk commented:
will merge this today after some doc tweaks
